presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.assertions;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+
+import java.util.List;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+class PlanMatchingVisitor
+        extends PlanVisitor<PlanMatchingContext, Boolean>
+{
+    private final Metadata metadata;
+    private final Session session;
+
+    public PlanMatchingVisitor(Session session, Metadata metadata)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    protected Boolean visitPlan(PlanNode node, PlanMatchingContext context)
+    {
+        List<PlanMatchingState> states = context.getPlanMatcher().matches(node, session, metadata, context.getSymbolAliases());
```

Comment:
The most problematic state is `anyTree()`. Lets take a look at below example.

Plan:
```
 - Project (a)
  - Filter (b)
    - Scan (c)
```

Pattern:
```
 - anyTree (A)
    - filter (B)
      - anyTree (C)
```

Matching first plan node (a) with `anyTree` (A) will return two states: 
 - (1) `anyTree` (A), as `anyTree` may match to any number of nodes
 - (2) `filter` (B)

Then matching second plan node (b) with pattern from state (1) will return one state:
 - (3) `anyTree` (A)
Then matching second plan node (b) with pattern from state (2) will return one state:
 - (4) `anyTree` (C)

Then matching third plan node (c) with pattern from state (3) will return one state:
 - (5) still `anyTree` (A)

Then matching third plan node (c) with pattern from state (4) will return:
 - (6) an empty (with no pattern) state

Note that (5) is not a termination state as it still contains child (`filter` pattern). State (6) is termination state as it contains no more patterns to check and we reached the end of plan.

Empty list of states means always no match.